### PR TITLE
Introduce the -dM option to print out the compiler-defined macros

### PR DESCRIPTION
**Pull request by pwisniewskimobica**
It was decided to add clang -dM option to flang.
It prints out all the macros defined - similarly as clang or gfortran does it.
It works only with llvm PR for this : https://github.com/flang-compiler/classic-flang-llvm-project/pull/90

### CODE REVIEW

File: tools/flang1/flang1exe/accpp.c
Comment (pawosm-arm) [inline tools/flang1/flang1exe/accpp.c:710]: `if<SPACE>(<condition>....`, `<SPACE>==<SPACE>`

File: tools/flang1/flang1exe/accpp.c
Comment (pawosm-arm) [inline tools/flang1/flang1exe/accpp.c:711]: same as above, add spaces around `==`

File: test/sema/dm_option.f90
Comment (pawosm-arm) [inline test/sema/dm_option.f90:6]: Add flags to ensure target system and architecture here.

File: tools/flang1/flang1exe/fpp.c
Comment (pawosm-arm) [inline tools/flang1/flang1exe/fpp.c:385]: see comments above for a similar piece of code.

**Review by pawosm-arm — CHANGES_REQUESTED** [review]
Better. But still needs some tweaking.

File: test/sema/dm_option.f90
Comment (pwisniewskimobica) [reply test/sema/dm_option.f90:12]: Below I paste the output of using this on my local machine- list is quite long.
@RichBarton-Arm mayby You have some idea what needs to be checked?

```
__GCC_ASM_FLAG_OUTPUTS__ : 1�
__code_model___ : 1�
__amd64__ : 1�
__amd64 : 1�
__x86_64 : 1�
__x86_64__ : 1�
__SEG_GS : 1�
__SEG_FS : 1�
__seg_gs : __attribute__�(�(�address_space�(�256�)�)�)�
__seg_fs : __attribute__�(�(�address_space�(�257�)�)�)�
__REGISTER_PREFIX__ : 
__NO_MATH_INLINES : 1�
__SSE2__ : 1�
__SSE2_MATH__ : 1�
__SSE__ : 1�
__SSE_MATH__ : 1�
__MMX__ : 1�
__GCC_HAVE_SYNC_COMPARE_AND_SWAP_1 : 1�
__GCC_HAVE_SYNC_COMPARE_AND_SWAP_2 : 1�
__GCC_HAVE_SYNC_COMPARE_AND_SWAP_4 : 1�
__GCC_HAVE_SYNC_COMPARE_AND_SWAP_8 : 1�
__SIZEOF_FLOAT128__ : 16�
unix : 1�
__unix : 1�
__unix__ : 1�
linux : 1�
__linux : 1�
__linux__ : 1�
__ELF__ : 1�
__gnu_linux__ : 1�
__FLOAT128__ : 1�
__SIZE_TYPE__ : long� unsigned� int�
__PTRDIFF_TYPE__ : long� int�
_LP64 : 1�
__LP64__ : 1�
__LONG_MAX__ : 9223372036854775807L�
__amd_64__amd64__ : 1�
__k8 : 1�
__k8__ : 1�
__THROW : 
__extension__ : 
__PGLLVM__ : 1�
__FLANG : 1
__STDC__ : 1
__FLANG_MAJOR__ : 99
__FLANG_MINOR__ : 99
__FLANG_PATCHLEVEL__ : 1
defined : `__GCC_ASM_FLAG_OUTPUTS__```

File: test/sema/dm_option.f90
Comment (pwisniewskimobica) [reply test/sema/dm_option.f90]: flags added

File: tools/flang1/flang1exe/accpp.c
Comment (pwisniewskimobica) [reply tools/flang1/flang1exe/accpp.c]: fixed

File: test/sema/dm_option.f90
Comment (pwisniewskimobica) [reply test/sema/dm_option.f90]: I decided to test target specific flags

File: tools/flang1/flang1exe/fpp.c
Comment (kiranchandramohan) [inline tools/flang1/flang1exe/fpp.c:379]: I am assuming accpp.c is the default pre-processor and hence by default only changes in accpp.c is only tested. Can you add tests for fpp.c also by appropriately switching on flags?

File: tools/flang1/flang1exe/accpp.c
Comment (kiranchandramohan) [inline tools/flang1/flang1exe/accpp.c:714]: Nit: I have not checked the code elsewhere, but do we use `printf` to usually display values? I would have assumed it will be `fprintf` or something like that.

File: tools/flang1/flang1exe/accpp.c
Comment (kiranchandramohan) [inline tools/flang1/flang1exe/accpp.c:710]: Nit: might be better to use `strncmp`.

File: test/sema/dm_option.f90
Comment (kiranchandramohan) [inline test/sema/dm_option.f90:1]: Technically this is a preprocessor test. Can we create a new sub-directory in test called preprocessor and add this test there? You might need to copy some files like `lit.local.cfg`.

File: tools/flang1/flang1exe/accpp.c
Comment (kiranchandramohan) [inline tools/flang1/flang1exe/accpp.c:940]: Nit: Why this check for incllist and not for others?

**Review by kiranchandramohan — COMMENTED** [review]
I have some minor comments. Please don't wait for my approval. Once @bryanpkc or @shivaramaarao approves, the PR can be submitted.

Please improve the commit message with the following details,
-> Mention in the commit message the addition of `list_macros` option to `flang1`.
-> Changes are made in the preprocessor related file `fpp.c` and `accpp.c` for printing the macros.

File: tools/flang1/flang1exe/accpp.c
Comment (pwisniewskimobica) [reply tools/flang1/flang1exe/accpp.c:940]: @kiranchandramohan i did it the same  as it is done in the rest file.Please look at lines ~1253

File: tools/flang1/flang1exe/accpp.c
Comment (pwisniewskimobica) [reply tools/flang1/flang1exe/accpp.c]: strncmp used
-->fixed

File: tools/flang1/flang1exe/accpp.c
Comment (pwisniewskimobica) [reply tools/flang1/flang1exe/accpp.c]: changed to `fprintf(gbl.outfil.....`
-->fixed

File: test/sema/dm_option.f90
Comment (pwisniewskimobica) [reply test/sema/dm_option.f90]: test location changed to `test/preprocessor/dm_option.f90`
-->fixed

File: tools/flang1/flang1exe/fpp.c
Comment (pwisniewskimobica) [reply tools/flang1/flang1exe/fpp.c:379]: tests for fpp module also added in dm_option.f90
for example `! Check amd64-linux architecture but from FPP module
! RUN: %flang -Hx,123,0x4000000 -dM -E -target x86_64-unknown-linux-gnu %s | FileCheck %s \
! RUN: --check-prefixes=CHECK-MACROS-AMD64,CHECK-MACROS-LINUX,CHECK-MACROS-COMMON-FPP`
-->fixed

**Comment by pwisniewskimobica** [comment]
@kiranchandramohan  all Your remarks applied.
Please take a look. 
Thanks

**Comment by bryanpkc** [comment]
Please reword the commit for correctness, e.g.
```
Introduce -dM option to print defined macros

Add flang1 option -list-macros, which is turned on by the driver option -dM.
Add code to fpp.c or accpp.c for printing the macros.
Add tests which checks macros from the FPP and ACCPP modules.
```

File: test/preprocessor/dm_option.f90
Comment (bryanpkc) [inline test/preprocessor/dm_option.f90:5]: I suggest moving the blank line from line 1 to just before this line, to separate the copyright header from the comment visually.

File: test/preprocessor/dm_option.f90
Comment (bryanpkc) [inline test/preprocessor/dm_option.f90:39]: Isn't this repeating the same test as line 31?

File: test/preprocessor/dm_option.f90
Comment (bryanpkc) [inline test/preprocessor/dm_option.f90:43]: Isn't this repeating the same test as line 35?

File: test/preprocessor/dm_option.f90
Comment (bryanpkc) [inline test/preprocessor/dm_option.f90:54]: Unnecessary blank comment.

File: tools/flang1/flang1exe/flgdf.h
Comment (bryanpkc) [inline tools/flang1/flang1exe/flgdf.h:75]: Add a comma after the word `FALSE`. This avoids future changes to this line when we add a new field below.

**Comment by pwisniewskimobica** [comment]
@bryanpkc All Your remarks applied.
Thank you

**Review by bryanpkc — APPROVED** [review]

**Pull request merged**
